Tribler/community/tunnel/hidden_community.py

```diff
@@ -257,25 +257,26 @@ def dht_callback(info_hash, peers, _):
 
                     self.create_key_request(info_hash, peer)
 
-        self._logger.debug("Doing dht lookup for hidden community")
+        self._logger.error("Doing dht lookup for hidden community")
         self.last_dht_lookup[info_hash] = time.time()
         self.dht_lookup(info_hash, dht_callback)
 
     def create_key_request(self, info_hash, sock_addr):
-        # 1. Select a circuit
-        self._logger.debug("Create key request: select circuit")
-        circuit = self.selection_strategy.select(None, self.hops[info_hash])
-        if not circuit:
-            self._logger.error("No circuit for key-request")
-            return False
-
-        # 2. Send a key-request message
-        self._logger.debug("Create key request: send key request")
-        cache = self.request_cache.add(KeyRequestCache(self, circuit, sock_addr, info_hash))
-        meta = self.get_meta_message(u'key-request')
-        message = meta.impl(distribution=(self.global_time,), payload=(cache.number, info_hash))
-        circuit.tunnel_data(sock_addr, TUNNEL_PREFIX + message.packet)
-        return True
+
+        def callback(circuit):
+            # Send a key-request message
+            self._logger.error("Create key request: send key request")
+            cache = self.request_cache.add(KeyRequestCache(self, circuit, sock_addr, info_hash))
+            meta = self.get_meta_message(u'key-request')
+            message = meta.impl(distribution=(self.global_time,), payload=(cache.number, info_hash))
+            circuit.tunnel_data(sock_addr, TUNNEL_PREFIX + message.packet)
+
+        # Create a circuit for getting the keys
+        self._logger.error("Create key request: create a circuit")
+        self.create_circuit(self.hops[info_hash],
```

Comment:
Correct. I discovered this method on a later moment, allowing dispersy data seems to be a cleaner option for this, as there are more places in the code where dispersy messages are tunneled to the end of a circuit (especially in the hiddencommunity I see lots of them). I'll have to clean up this code a bit. :)